File: src/snippets/guards/quick-start/strands-agents-py/Wrap.py
import os

from strands import Agent, tool
from arcjet.guard import LocalDetectSensitiveInfo, launch_arcjet
from arcjet.guard.strands_agents import guard_hooks, guard_tool
from arcjet_sensitive_info_rampart import rampart

# Create one Arcjet client and reuse it across agent runs. Rampart
# detects bank account and routing numbers locally.
arcjet = launch_arcjet(
    key=os.environ["ARCJET_KEY"],
    sensitive_info_backend=rampart(),
)
detect_pii = LocalDetectSensitiveInfo(
    deny=["BANK_ACCOUNT", "ROUTING_NUMBER"],
)


def email_tools(user):
    @tool
    def get_client_record() -> dict:
        """Get the account details on file for the current customer."""
        return user.record

    @tool
    def send_email(recipient: str, body: str) -> str:
        """Send an email."""
        email_provider.send(to=recipient, body=body)
        return "sent"

    # This adapter submits action and rules. The recipient
    # allow list is not enforced in this sample.
    guarded_send_email = guard_tool(
        guard=arcjet,
        tool=send_email,
        action="email.sent",
        rules=lambda arguments: [detect_pii(arguments["body"])],
    )

    return get_client_record, guarded_send_email


def run_email_agent(user, prompt: str):
    get_client_record, send_email = email_tools(user)
    agent = Agent(
        tools=[get_client_record, send_email],
        hooks=[
            guard_hooks(guard=arcjet, session_id=user.session_id)
        ],
    )
    return agent(prompt)

File: src/snippets/guards/quick-start/claude-agent-sdk/js/Wrap.ts
import {
  createSdkMcpServer,
  query,
  tool,
} from "@anthropic-ai/claude-agent-sdk";
import { launchArcjet, localDetectSensitiveInfo } from "@arcjet/guard";
import { rampart } from "@arcjet/sensitive-info-rampart";
import { guardHooks, guardTool } from "@arcjet/guard/claude-agent-sdk/v0";
import { z } from "zod";

// Create one Arcjet client and reuse it across agent runs. Rampart
// detects bank account and routing numbers locally.
const arcjet = launchArcjet({
  key: process.env.ARCJET_KEY!,
  sensitiveInfoBackend: rampart(),
});
const detectPii = localDetectSensitiveInfo({
  deny: ["BANK_ACCOUNT", "ROUTING_NUMBER"],
});

export function emailTools(user: {
  record: {
    name: string;
    bankAccount: string;
    routingNumber: string;
  };
}) {
  const getClientRecord = tool(
    "get_client_record",
    "Get the account details on file for the current customer",
    {},
    async () => ({
      content: [{ type: "text", text: JSON.stringify(user.record) }],
    }),
  );

  // This adapter accepts action and rules. It doesn't accept
  // inputs.
  const sendEmail = guardTool(
    arcjet,
    tool(
      "send_email",
      "Send an email",
      {
        recipient: z.string(),
        body: z.string(),
      },
      async ({ recipient, body }) => {
        await emailProvider.send({ to: recipient, body });
        return {
          content: [{ type: "text", text: "sent" }],
        };
      },
    ),
    {
      action: "email.sent",
      rules: ({ body }) => [detectPii(body)],
    },
  );

  return { getClientRecord, sendEmail };
}

export async function runEmailAgent(
  user: {
    record: {
      name: string;
      bankAccount: string;
      routingNumber: string;
    };
  },
  sessionId: string,
  prompt: string,
) {
  const { getClientRecord, sendEmail } = emailTools(user);
  const server = createSdkMcpServer({
    name: "email",
    version: "1.0.0",
    tools: [getClientRecord, sendEmail],
  });

  for await (const message of query({
    prompt,
    options: {
      sessionId,
      mcpServers: { email: server },
      allowedTools: ["mcp__email__get_client_record", "mcp__email__send_email"],
      hooks: guardHooks(arcjet, { sessionId }),
    },
  })) {
    if (message.type === "result" && message.subtype === "success") {
      return message.result;
    }
  }
}
